index.rst

```diff
@@ -67,6 +71,40 @@ Alternative designs
 
+Replace ``Cookie`` and ``Authorization`` headers
+------------------------------------------------
+
+While the ``auth_request`` handler cannot remove headers, NGINX and the Kubernetes ``Ingress`` annotations do provide a way of replacing headers in the request with headers returned by the ``auth_request`` handler before passing them to the backend.
+We could make use of this by having Gafaelfawr return a ``Cookie`` header that is the same as the incoming request but Gafaelfawr cookie stripped out of it, and then configure the ingress to replace the incoming header with that header via:
+
+.. code-block:: yaml
+
+   nginx.ingress.kubernetes.io/auth-response-headers: Cookie
+
+(plus any additional headers from Gafaelfawr that the protected application uses).
+A simpler approach would work for the ``Authorization`` header: listing the header in ``auth-response-headers`` without sending that header in the ``auth_request`` handler response should strip the header from the request before sending it to the backend service.
+This would only work for services that do not reuse the ``Authorization`` header for their own purposes (JupyterHub does this).
+
+For backend services that must receive a token in the ``Authorization`` header (CADC's TAP service is currently in this category), Gafaelfawr could be configured to return a delegated token in the ``Authorization`` header.
+
+Advantages:
+
+- Works transparently with the current Rubin Science Platform design, with no changes required to protected services, routes, or hostnames.
+- Addresses both the cookie and ``Authorization`` header cases.
+- Conceptually simple, and collects all of our request manipulation code in the same place (the Gafaelfawr ``auth_request`` handler).
+
+Disadvantages:
+
+- Requires parsing and surgery on the ``Cookie`` header in Gafaelfawr, which opens the possibility of unexpected browser interactions or problems with invalid but still working headers.
+- Only protects against token leakage to authenticated URLs that are protected by Gafaelfawr.
+  Requests that do not go through an ``auth_request`` handler will still leak cookies and ``Authorization`` headers.
+  This includes any request to a backend service that uses Gafaelfawr's OpenID Connect support instead.
+  This could be mitigated by supporting a stripped-down ``auth_request`` handler mode that only cleans the headers and sending all requests, even unauthenticated requests, through an ``auth_request`` handler, at the cost of additional complexity and possible fragility.
+- This adds additional complexity to the required ingress configuration to use Gafaelfawr, which is already very complex.
+  If we take this approach, it may be worth adding Gafaelfawr support for a custom ingress resource and have Gafaelfawr add the appropriate annotations and generate the real ``Ingress` resource from that custom resource.
```

Comment:
```suggestion
  If we take this approach, it may be worth adding Gafaelfawr support for a custom ingress resource and have Gafaelfawr add the appropriate annotations and generate the real ``Ingress`` resource from that custom resource.
```